### PR TITLE
Source maps: allow to specify that an expression has no debug info in text parsers

### PR DESCRIPTION
**Pull request by vouillon**
The comment `;;@` with nothing else can be used to specify that the following expression does not have any debug info associated to it. This can be used to stop the automatic propagation of debug info in the text parsers.

The text printer has also been updated to output this comment when needed.

### CODE REVIEW

**Review by tlively — COMMENTED** [review]
Isn't 0,0,0 a valid debug location? I think it would be better to use an empty `std::optional` to signify the absence of a debug location.

File: src/parser/contexts.h [hunk 1]
```diff
@@ -1759,7 +1764,8 @@ struct ParseDefsCtx : TypeParserCtx<ParseDefsCtx> {
       assert(wasm.debugInfoFileNames.size() == it->second);
       wasm.debugInfoFileNames.push_back(std::string(file));
     }
-    irBuilder.setDebugLocation({it->second, *line, *col});
+    irBuilder.setDebugLocation(
+      Function::DebugLocation({it->second, *line, *col}));
```
Comment (kripken) [inline src/parser/contexts.h:1768]: Why is this needed?

File: src/passes/Print.cpp [hunk 2]
```diff
@@ -117,7 +117,7 @@ struct PrintSExpression : public UnifiedExpressionVisitor<PrintSExpression> {
 
   Module* currModule = nullptr;
   Function* currFunction = nullptr;
-  Function::DebugLocation lastPrintedLocation;
+  std::optional<Function::DebugLocation> lastPrintedLocation;
```
Comment (kripken) [inline src/passes/Print.cpp:120]: Maybe add a comment here about what the empty option means.

File: src/wasm/wasm-binary.cpp [hunk 3]
```diff
@@ -2686,12 +2686,11 @@ void WasmBinaryReader::readFunctions() {
 
     readVars();
 
-    std::swap(func->prologLocation, debugLocation);
+    func->prologLocation = debugLocation;
     {
       // process the function body
       BYN_TRACE("processing function: " << i << std::endl);
       nextLabel = 0;
-      debugLocation.clear();
```
Comment (kripken) [inline src/wasm/wasm-binary.cpp:2694]: Why is this no longer needed?

File: src/wasm-ir-builder.h [hunk 4]
```diff
@@ -238,7 +238,7 @@ class IRBuilder : public UnifiedExpressionVisitor<IRBuilder, Result<>> {
   Module& wasm;
   Function* func;
   Builder builder;
-  std::optional<Function::DebugLocation> debugLoc;
+  std::optional<std::optional<Function::DebugLocation>> debugLoc;
```
Comment (kripken) [inline src/wasm-ir-builder.h:241]: Worth a comment here too I think.

File: src/wasm/wasm-s-parser.cpp [hunk 5]
```diff
@@ -219,6 +219,10 @@ void SExpressionParser::parseDebugLocation() {
   while (debugLocEnd[0] && debugLocEnd[0] != '\n') {
     debugLocEnd++;
   }
+  if (debugLocEnd == debugLoc) {
+    loc = nullptr;
+    return;
+  }
```
Comment (kripken) [inline src/wasm/wasm-s-parser.cpp:225]: How does this relate to the PR - is it necessary because of something, or necessary for a new test, or something else?

File: src/parser/contexts.h (hunk 1 above)
Comment (vouillon) [reply src/parser/contexts.h:1768]: `setDebugLocation` has been changed to take a `std::optional<Function::DebugLocation>`, so the previous code no longer compiles.

File: src/wasm/wasm-s-parser.cpp (hunk 5 above)
Comment (vouillon) [reply src/wasm/wasm-s-parser.cpp:225]: This is just parsing `;;@` with nothing after.

File: src/wasm/wasm-binary.cpp (hunk 3 above)
Comment (vouillon) [reply src/wasm/wasm-binary.cpp:2694]: This change exposed a bug. If the first segment starts a bit before the beginning of the function, it may still cover some instructions. So, one should not clear the `debugLocation` here.

The `fn_prolog_epilog.debugInfo` test is failing without this change, since the debug information associated to the nop instruction at the beginning of the function is dropped.

Comment (kripken) [reply src/wasm/wasm-binary.cpp:2694]: Which test do you mean by `fn_prolog_epilog.debugInfo`? I grepped for that name and didn't find anything.

Comment (vouillon) [reply src/wasm/wasm-binary.cpp:2694]: There is a file `test/fn_prolog_epilog.debugInfo.wasm`.

File: src/passes/Print.cpp [hunk 6]
```diff
@@ -117,7 +117,12 @@ struct PrintSExpression : public UnifiedExpressionVisitor<PrintSExpression> {
 
   Module* currModule = nullptr;
   Function* currFunction = nullptr;
-  Function::DebugLocation lastPrintedLocation;
+  // Keep track of the last printed debug location to avoid printing
+  // repeated debug locations for children. nullopt means that we have
+  // not yet printed any debug location, or that we last printed an
+  // annotation indicating that the expression had not associated
```
Comment (kripken) [inline src/passes/Print.cpp:123]:
```suggestion
  // annotation indicating that the expression had no associated
```

File: test/fib-dbg.wasm.fromBinary [hunk 7]
```diff
@@ -209,6 +209,7 @@
      (br $label$4)
     )
    )
+   ;;@
```
Comment (kripken) [inline test/fib-dbg.wasm.fromBinary:212]: I tried to look for this in the binary. What I did is add a printf here:

https://github.com/WebAssembly/binaryen/blob/f44dcd4482f5bbb0fb5b567de4a9425ed949c939/src/wasm/wasm-binary.cpp#L2965

I believe that is the right place for reading a source map entry that says "this has no debug info, explicitly". But I don't see it reached. Am I doing something wrong?

Comment (vouillon) [reply test/fib-dbg.wasm.fromBinary:212]: No, you are right. One should probably not clear the debug location here:
https://github.com/WebAssembly/binaryen/blob/f44dcd4482f5bbb0fb5b567de4a9425ed949c939/src/wasm/wasm-binary.cpp#L2933

**Comment by vouillon** [comment]
I have opened a separate PR #6550 that include the source map fixes. I'll rebase this PR once the other is merged.

**Comment by kripken** [comment]
Is this ready for review after the other PR landed?

**Comment by vouillon** [comment]
Yes, this is ready

File: test/lit/debug/source-map-stop.wast [hunk 8]
```diff
@@ -124,4 +124,41 @@
...
     (i32.const 2)
   )
+
+  ;; CHECK:      (func $foo (param $x i32) (param $y i32)
+  ;; CHECK-NEXT:  ;;@ src.cpp:90:1
+  ;; CHECK-NEXT:  (if
+  ;; CHECK-NEXT:   ;;@
+  ;; CHECK-NEXT:   (i32.add
+  ;; CHECK-NEXT:    (local.get $x)
+  ;; CHECK-NEXT:    (local.get $y)
```
Comment (kripken) [inline test/lit/debug/source-map-stop.wast:134]: To make sure I understand, these `local.get` have no debug annotation, like their parent, and so we did not print anything for them? That is, "nothing" propagates to children just like an actual debug location does?

Comment (kripken) [reply test/lit/debug/source-map-stop.wast:134]: This might be worth clarifying in the README.

Comment (vouillon) [reply test/lit/debug/source-map-stop.wast:134]: Right. I have updated the README.

**Review by kripken — APPROVED** [review]
